build with gcc-6.3.0/cuda 9

Creating issue and copying relevant discussion from #35

From Dhyan:

Building without CUDA / kokkos and with gcc-6.3.0 gives errors like
===========================================================
/fusion/usc/opt/gcc/gcc-6.3.0/include/c++/6.3.0/bits/unique_ptr.h:158:17: note:   candidate expects 0 arguments, 1 provided
In file included from /home/nathd/omega_h/src/Omega_h_array.hpp:4:0,
                 from /home/nathd/omega_h/src/Omega_h_comm.hpp:7,
                 from /home/nathd/omega_h/src/Omega_h_array_ops.hpp:6,
                 from /home/nathd/omega_h/src/Omega_h_rcFields.cpp:4:
/home/nathd/omega_h/src/Omega_h_defines.hpp: In instantiation of ‘auto Omega_h::apply_to_omega_h_types(Omega_h_Type, const F&) [with F = Omega_h::Mesh::get_rc_mesh_tag_from_rc_tag(Omega_h::Int, const Omega_h::TagBase*)::<lambda(auto:1)>]’:
/home/nathd/omega_h/src/Omega_h_rcFields.cpp:340:4:   required from here
/home/nathd/omega_h/src/Omega_h_defines.hpp:93:15: error: use of ‘Omega_h::Mesh::get_rc_mesh_tag_from_rc_tag(Omega_h::Int, const Omega_h::TagBase*)::<lambda(auto:1)> [with auto:1 = int]’ before deduction of ‘auto’
       return f(I32{});
              ~^~~~~~~
/home/nathd/omega_h/src/Omega_h_defines.hpp:93:21: error: return-statement with a value, in function returning 'void' [-fpermissive]
       return f(I32{});
                     ^
/home/nathd/omega_h/src/Omega_h_defines.hpp:97:15: error: use of ‘Omega_h::Mesh::get_rc_mesh_tag_from_rc_tag(Omega_h::Int, const Omega_h::TagBase*)::<lambda(auto:1)> [with auto:1 = long int]’ before deduction of ‘auto’
       return f(I64{});

=====================================================================

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,6 +1,6 @@
 cmake_minimum_required(VERSION 3.18.0...${CMAKE_VERSION})
 
-project(Omega_h VERSION 10.4.0 LANGUAGES CXX)
+project(Omega_h VERSION 10.5.0 LANGUAGES CXX)
 set(Omega_h_USE_STK_DEFAULT OFF)
 option(Omega_h_USE_STK "Whether to build the STK interface" ${Omega_h_USE_STK_DEFAULT})
 

--- a/src/Omega_h_defines.hpp
+++ b/src/Omega_h_defines.hpp
@@ -83,7 +83,7 @@ typedef I64 GO;
 typedef double Real;
 
 template <typename F>
-auto apply_to_omega_h_types(Omega_h_Type type, const F& f) {
+auto apply_to_omega_h_types(Omega_h_Type type, const F&& f) {
   switch (type) {
     case OMEGA_H_I8: {
       return f(I8{});

--- a/src/Omega_h_file.cpp
+++ b/src/Omega_h_file.cpp
@@ -293,7 +293,7 @@ static void write_tag(std::ostream& stream, TagBase const* tag,
     using T = decltype(type);
     write_array(stream, as<T>(tag)->array(), is_compressed, needs_swapping);
   };
-  apply_to_omega_h_types(tag->type(), f);
+  apply_to_omega_h_types(tag->type(), std::move(f));
 }
 static void write_rc_tag(std::ostream& stream, TagBase const* tag,
     Int ent_dim, Mesh *mesh, bool is_compressed, bool needs_swapping) {
@@ -343,7 +343,7 @@ static void read_tag(std::istream& stream, Mesh* mesh, Int d,
       mesh->add_tag(d, name, ncomps, array, true);
     }
   };
-  apply_to_omega_h_types(static_cast<Omega_h_Type>(type), f);
+  apply_to_omega_h_types(static_cast<Omega_h_Type>(type), std::move(f));
 }
 
 static void write_sets(std::ostream& stream, Mesh* mesh, bool needs_swapping) {

--- a/src/Omega_h_rcFields.cpp
+++ b/src/Omega_h_rcFields.cpp
@@ -224,11 +224,11 @@ template <typename T>
 void Mesh::set_rc_from_mesh_array(Int ent_dim, Int ncomps, LOs class_ids,
     std::string const& name, Read<T> array) {
   OMEGA_H_TIME_FUNCTION;
-  auto tag_itr = rc_tag_iter(ent_dim,name);
+  auto tag_itr = rc_tag_iter(ent_dim, name);
   auto b_field =
       get_rc_array_from_mesh_array(ent_dim, ncomps, name, class_ids, array);
-  if(!tag_itr.had_tag) {
-    add_rcField<T>(class_ids,ent_dim,name,ncomps);
+  if (!tag_itr.had_tag) {
+    add_rcField<T>(class_ids, ent_dim, name, ncomps);
   }
   set_rcField_array(ent_dim, name, b_field);
 }
@@ -333,11 +333,11 @@ std::unique_ptr<Tag<T>> Mesh::get_rc_mesh_tag_from_rc_tag(
 
 std::unique_ptr<TagBase> Mesh::get_rc_mesh_tag_from_rc_tag(
     Int ent_dim, TagBase const* tag) {
-  auto new_tag = apply_to_omega_h_types(tag->type(), [&](auto t) {
-    using T = decltype(t);
-    return std::unique_ptr<TagBase>{
-        get_rc_mesh_tag_from_rc_tag(ent_dim, as<T>(tag))};
-  });
+  auto new_tag = apply_to_omega_h_types(
+      tag->type(), [&,this](auto t) -> std::unique_ptr<TagBase> {
+        using T = decltype(t);
+        return this->get_rc_mesh_tag_from_rc_tag(ent_dim, as<T>(tag));
+      });
   return new_tag;
 }
 
@@ -391,7 +391,7 @@ bool Mesh::has_rcField(Int ent_dim, std::string const& name) const {
 
 void Mesh::remove_rcField(Int ent_dim, std::string const& name) {
   auto new_name = get_rc_name(name);
-  auto tag_itr= rc_tag_iter(ent_dim, new_name);
+  auto tag_itr = rc_tag_iter(ent_dim, new_name);
   if (tag_itr.had_tag) {
     rc_field_tags_[ent_dim].erase(tag_itr.it);
   }
@@ -414,21 +414,20 @@ void Mesh::set_rcField_array(
 void Mesh::reduce_rcField(Int ent_dim, std::string const& name, Omega_h_Op op) {
   auto new_name = get_rc_name(name);
 
-  auto tag_itr= rc_tag_iter(ent_dim, new_name);
+  auto tag_itr = rc_tag_iter(ent_dim, new_name);
   OMEGA_H_CHECK(tag_itr.had_tag);
   // auto tagbase
   auto* tagbase = tag_itr.it->get();
   const auto class_ids = tagbase->class_ids();
   const auto ncomps = tagbase->ncomps();
 
-  auto f = [&](auto t) {
+  apply_to_omega_h_types(tagbase->type(), [&, this](auto t) {
     using T = decltype(t);
     auto mesh_array =
-        get_rc_mesh_array<T>(ent_dim, ncomps, new_name, class_ids);
-    auto out = reduce_array(ent_dim, mesh_array, ncomps, op);
-    set_rc_from_mesh_array(ent_dim, ncomps, class_ids, new_name, out);
-  };
-  apply_to_omega_h_types(tagbase->type(), f);
+        this->get_rc_mesh_array<T>(ent_dim, ncomps, new_name, class_ids);
+    auto out = this->reduce_array(ent_dim, mesh_array, ncomps, op);
+    this->set_rc_from_mesh_array(ent_dim, ncomps, class_ids, new_name, out);
+  });
 }
 
 void Mesh::sync_rcField(Int ent_dim, std::string const& name) {
@@ -438,14 +437,13 @@ void Mesh::sync_rcField(Int ent_dim, std::string const& name) {
 
   const auto ncomps = (*tag_itr.it)->ncomps();
   const auto class_ids = (*tag_itr.it)->class_ids();
-  auto f = [&](auto t) {
+  apply_to_omega_h_types((*tag_itr.it)->type(), [&, this](auto t) {
     using T = decltype(t);
     auto mesh_array =
-        get_rc_mesh_array<T>(ent_dim, ncomps, new_name, class_ids);
-    auto out = sync_array(ent_dim, mesh_array, ncomps);
-    set_rc_from_mesh_array(ent_dim, ncomps, class_ids, new_name, out);
-  };
-  apply_to_omega_h_types((*tag_itr.it)->type(), f);
+        this->get_rc_mesh_array<T>(ent_dim, ncomps, new_name, class_ids);
+    auto out = this->sync_array(ent_dim, mesh_array, ncomps);
+    this->set_rc_from_mesh_array(ent_dim, ncomps, class_ids, new_name, out);
+  });
 }
 
 bool Mesh::change_all_rcFieldsToMesh() {
@@ -458,13 +456,12 @@ bool Mesh::change_all_rcFieldsToMesh() {
       auto const& name = rc_tag->name();
       auto ncomps = rc_tag->ncomps();
       auto class_ids = rc_tag->class_ids();
-      auto f = [&](auto t) {
+      apply_to_omega_h_types(rc_tag->type(), [&, this](auto t) {
         using T = decltype(t);
         // TODO can change_tagToMesh just take a tag? because we get all the
         // data from the tag anyways...
-        change_tagToMesh<T>(ent_dim, ncomps, name, class_ids, false);
-      };
-      apply_to_omega_h_types(rc_tag->type(), f);
+        this->change_tagToMesh<T>(ent_dim, ncomps, name, class_ids, false);
+      });
       OMEGA_H_CHECK(has_tag(ent_dim, name));
     }
     rc_field_tags_[ent_dim].clear();
@@ -484,20 +481,17 @@ bool Mesh::change_all_rcFieldsTorc() {
         auto const& name = tag->name();
         auto ncomps = tag->ncomps();
         auto class_ids = tag->class_ids();
-        auto f = [&](auto t) {
+        apply_to_omega_h_types(tag->type(), [&, this](auto t) {
           using T = decltype(t);
           // TODO can change_tagToMesh just take a tag? because we get all the
           // data from the tag anyways...
-          change_tagTorc<T>(ent_dim, ncomps, name, class_ids, false);
-        };
-        apply_to_omega_h_types(tag->type(), f);
+          this->change_tagTorc<T>(ent_dim, ncomps, name, class_ids, false);
+        });
       }
     }
     tags_[ent_dim].erase(
         std::remove_if(tags_[ent_dim].begin(), tags_[ent_dim].end(),
-            [](const auto& tag) {
-              return is_rc_tag(tag->name());
-            }),
+            [](const auto& tag) { return is_rc_tag(tag->name()); }),
         tags_[ent_dim].end());
   }
   return changed;

